### ISSUE TITLE
Search in book(s)

### ISSUE THREAD

**Issue by arthurattwell** [electricbookworks/electric-book#61]
As a site user, I want to be able to search the entire site for phrases, so that I can find what I'm looking for instantly.

**Comment by arthurattwell** [comment]
Migrated from Trello. Key comments from there:

**Arthur Attwell, Nov 15, 2016 at 11:00 AM**
This is working from template v0.5.0. @maxbarners working on improving results, e.g. better results when search string includes multiple words, and removing unwanted (e.g. 'Null') results.

For testing, check out [electric-book-workflow](https://github.com/electricbookworks/electric-book-workflow) and [lawful-living](http://fireandlion.github.io/lawful-living).

**Steve Barnett, Nov 22, 2016 at 9:43 AM**
@arthurattwell ohai!

I've been fiddling with this and have a PR in progress: [electricbookworks/electric-book: Pull Request 32](https://github.com/electricbookworks/electric-book/pull/32) (#32). There are two fiddly issues that I'd like your thoughts on, please.

1. Stemming

Elasticlunr stems search terms, but we use the full search term for highlighting. That means that viewing a [search result for encoding](http://electricbook.works/guide/text/20-poetry.html?query=encoding) has marked text, but a [search result for encodings](http://electricbook.works/guide/text/20-poetry.html?query=encodings) doesn't (even though it [correctly appears in the search results](http://electricbook.works/guide/text/search.html?query=encodings))

It looks like [we could fudge this](http://stackoverflow.com/questions/28346291/lunr-return-the-stem-of-the-searched-terms-so-i-can-highlight-it-within-results) to make it work.

Is it worth spending the time on?

2. Multi-word searches

Elasticlunr tokenises the search term, so [basic poetry](http://electricbook.works/guide/text/20-poetry.html?query=basic%20poetry) gets you basic and poetry.

Are we happy with this behaviour?

**Steve Barnett, Nov 22, 2016 at 9:47 AM**
Erm, another question: I want to improve the search results, but I need some kind of nudge / clarification. What will be useful to people here? Excerpt of the result page? Headings from the chapter?

Elasticlunr doesn't give us much back, it seems, but we do have the whole page in the datastore.

**Arthur Attwell, Nov 22, 2016 at 11:30 AM**
@maxbarners Cool, some replies.

1. Stemming

   I suggest we temporarily set the highlighing colour to inherit (in the Sass config), effectively turning it off. Then we revisit later, doing the fudge. I'd rather spend time now on better snippets in search results. (More on that below.)

2. Multi-word searches

   Ah, this one is tricky. What will users most expect? I suspect phrases should match as a higher priority. Can we chat about this? Hoping you can tell me more about how elasticlunr's [tokenizer](http://elasticlunr.com/docs/tokenizer.js.html) and seperator [sic] options work.

3. Search results snippets

   In a perfect/magical world, the search results would give users the snippet of text that includes the query string, from a few words before the query string to a few words after. Is there a way to get that from the datastore? 

   Second prize might be the heading and opening paragraph of the chapter?

PS If at any point you feel we'll get what we need way faster with server-based search (even a paid service), I'm totally open to that. I can imagine the template one day including a config option to switch between elasticlunr, which ships with the template, and a given server-based service that we like. This might even tie into the potential business model behind hosting Electric Book instances, where hosts can choose to offer more powerful, server-based search for clients who want it. Am I getting distracted now? Yes, almost certainly.

**Steve Barnett, Nov 22, 2016 at 3:24 PM**
![image](https://cloud.githubusercontent.com/assets/1053331/21718373/ae895c62-d41f-11e6-9c9c-5a339b0732f2.png)

**Steve Barnett, Nov 22, 2016 at 3:26 PM**
@arthurattwell I've poked at it a bit more. :)
Using some fancy filtering, I've managed to pluck out the first paragraph as an excerpt. (`post.excerpt` is a thing, but `page.excerpt` isn't, apparently `¯\_(ツ)_/¯ `).

I've added some headers for structure.

I think it needs a `&hellip;`-like: what do you think would be best? Also, we could alternatively use `truncatewords:20` instead of the first `<p>`.

**Steve Barnett, Nov 22, 2016 at 3:30 PM**
Record-keeping: Arthur and Steve discussed the options, and how the client-side search is getting fiddly, fast. It was agreed to do a cheap and cheerful one for now, and think later about doing the external, server-side, one mentioned earlier in this thread.

**Arthur Attwell, Nov 22, 2016 at 3:30 PM**
@maxbarners Great! I'm thinking &hellip; and truncatewords: 20 is a good way to go.

**Steve Barnett, Nov 22, 2016 at 3:34 PM**
Done: [https://github.com/electricbookworks/electric-book/pull/32/commits/fff2a30bb99534754a30c18f00e763ad8dd0f763](https://github.com/electricbookworks/electric-book/pull/32/commits/fff2a30bb99534754a30c18f00e763ad8dd0f763).

`truncatewords` uses three `.`s instead of an ellipsis by default. [scream]

**Arthur Attwell, Dec 14, 2016 at 1:30 PM**
@maxbarners Client-side search, with elasticlunr, is now working fine in the EB template out of the box. So the next question is: are we happy that that's it for now, or do we want to bake in a more powerful server-side option?

**Steve Barnett, Dec 14, 2016 at 5:13 PM**
@arthurattwell I think we should look / think seriously about server-side search, for extra power but also for weight and modularity.

*Weight*

The search index is quite chunky, and I suspect will be very chunky for big books, especially on small screens / phones. Moving that weight to a server and just to-and-fro-ing the result set (of data) would be a big win for me.

*Modularity*

I have been thinking about how maybe EB should be quite small, and not try and do All The Things out of the box. Maybe a quick prioritised list of what EB needs (must, should, could?), with a mark for use cases / who would be interested in it (and use it!).

**Comment by SteveBarnett** [comment]
@arthurattwell Are we happy that client-side search is done?
If yes, are we keeping this issue open for server-side search, or starting a fresh issue? :)

**Comment by arthurattwell** [comment]
@SteveBarnett I think we should update/tweak the template's default exclusions, based on what we found on another project where a page on 'copyright' was being excluded.

Does it feel odd to you to be hard-coding those exclusions into `search-data.js`? Should we be abstracting the list to some kind of config file, like `meta.yml`, so it's more intuitive to find and edit, and safer from being overwritten during upgrades?

I've added issue #93 for server-side search.
